bin/jsii-release-golang

```diff
@@ -56,68 +72,121 @@ if [ -z "${GITHUB_TOKEN:-}" ]; then
+  owner=$(echo ${full_module_name} | cut -d '/' -f 2)
+  repo=$(echo ${full_module_name} | cut -d '/' -f 3)
+  echo ${owner}/${repo}
+}
+
+# Creates the necessary tag for a given module.
+# If the module is located at the root directory, the tag will in the form of `v${module_version}`.
+# Otherwise, the module name is prefixed to it, e.g `my-module/v${module_version}`.
+function create_tag {
+  module=$1
+
+  module_name=$(basename ${module})
+  module_version=$(extract_version ${module})
+
+  if [ ${module} = ${dir} ]; then
+    # root module
+    tag_name=v${module_version}
+  else
+    # submodule
+    tag_name=${module_name}/v${module_version}
+  fi
 
-unique_versions=($(echo "${versions[@]}" | tr ' ' '\n' | sort -u | tr '\n' ' '))
-count_unique_version=$(echo "${#unique_versions[@]}")
+  git tag -a ${tag_name} -m ${tag_name}
+  if $dry_run; then
+    echo "Will push tag: ${tag_name}"
+  else
+    git push origin ${tag_name}
+  fi
+}
+
+# Delete all modules in the target repository in preparation of
+# syncing it with the source code located in `DIR`.
+function purge_modules {
+  repo_dir=$1
+
+  # first remove all submodules
+  for subdir in $(find ${repo_dir} -mindepth 1 -maxdepth 1 -type d); do
+    if [ -f ${subdir}/go.mod ]; then
+      rm -r ${subdir}
+    fi
+  done
+
+  # if the root is a module too, remove the entire repo.
+  if [ -f ${dir}/go.mod ]; then
+    rm -r ${repo_dir}/*
+  fi
 
-if [ ${count_unique_version} -gt 1 ]; then
-  error "multiple versions are not supported. please use a single version across all your modules."
-fi
+}
 
-if [ -z "${VERSION:-}" ]; then
+modules=()
 
-  if [ ${count_unique_version} -eq 0 ]; then
-    error "No version files found. Please supply a version using the VERSION env variable."
-  fi
+# collect mod files from root and subdirectories
+for go_mod in $(find ${dir} -name go.mod -mindepth 1 -maxdepth 2); do
+  modules+=($(dirname ${go_mod}))
+done
 
-  # must be a single element at this point
-  VERSION=${unique_versions[0]}
+# at least one go module must exist
+if [ ${#modules[@]} -eq 0 ]; then
+  error "No go modules detected in ${dir}"
 fi
 
-GIT_COMMIT_MESSAGE=${GIT_COMMIT_MESSAGE:-"chore(release): ${VERSION}"}
-
-module_name=$(grep module ${go_mod} | cut -d ' ' -f 2)
-domain=$(echo ${module_name} | cut -d '/' -f 1)
+echo "Detected modules:"
+for module in "${modules[@]}"; do
+  echo "  ${module}"
+done
 
-if [ ${domain} != "github.com" ]; then
-  # the way we clone the repo with the token is a GitHub thing.
-  error "'${domain}' domain is not supported. Only GitHub repositories are allowed"
-fi
 
-if [ -z "${GITHUB_REPO:-}" ]; then
-  owner=$(echo ${module_name} | cut -d '/' -f 2)
-  repo=$(echo ${module_name} | cut -d '/' -f 3)
-  GITHUB_REPO=${owner}/${repo}
-fi
+GIT_COMMIT_MESSAGE=${GIT_COMMIT_MESSAGE:-$(create_release_message ${modules})}
+GITHUB_REPO=${GITHUB_REPO:-$(extract_repo ${modules})}
 
-target_repo_dir=$(mktemp -d)/repo
+target_repo_dir=$(mktemp -d)/$(basename ${GITHUB_REPO})
```

Comment:
why no just `repo`?

Reply:
Was used to detect top level modules, but I switched the logic so its no longer necessary.